File: sdk/typescript/src/session.ts
import { OctoberBusAdminClient, OctoberBusClient, OctoberBusScopeClient } from './client.js'
import { BusError } from './errors.js'
import type {
  Agent,
  AgentLifecycle,
  BusMessage,
  BusTask,
  RegisterAgentInput,
  RegisterAgentResult
} from './protocol.js'

export interface AgentSessionOptions {
  address: string
  scopeToken: string
  registration: RegisterAgentInput
  heartbeatIntervalMs?: number
  initialLifecycle?: AgentLifecycle
  initialReady?: boolean
  signal?: AbortSignal
}

export interface InboxPollingOptions {
  limit?: number
  waitMs?: number
  signal?: AbortSignal
}

export interface ClaimedTaskResult<T> {
  task: BusTask
  value: T
}

export function requiredEnvironmentValue(
  environment: Readonly<Record<string, string | undefined>>,
  name: string
): string {
  const value = environment[name]
  if (!value) throw new Error(`${name} is required`)
  return value
}

export function newIdempotencyKey(): string {
  return crypto.randomUUID()
}

export class OctoberBusAgentSession {
  readonly client: OctoberBusClient
  readonly registration: RegisterAgentResult
  readonly done: Promise<void>

  private readonly leaseMs: number
  private readonly heartbeatIntervalMs: number
  private readonly signal: AbortSignal | undefined
  private readonly lifecycleAbort = new AbortController()
  private readonly abortListener = (): void => { void this.close() }
  private lifecycle: AgentLifecycle
  private ready: boolean
  private timer: ReturnType<typeof setTimeout> | undefined
  private lifecycleQueue: Promise<void> = Promise.resolve()
  private closePromise: Promise<void> | undefined
  private resolveDone!: () => void
  private closed = false
  private sessionError: unknown

  private constructor(options: AgentSessionOptions, registration: RegisterAgentResult) {
    this.registration = registration
    this.client = new OctoberBusClient(options.address, registration.agentToken)
    this.leaseMs = options.registration.leaseMs || 300_000
    this.heartbeatIntervalMs = options.heartbeatIntervalMs ?? Math.floor(this.leaseMs / 3)
    this.lifecycle = options.initialLifecycle ?? 'starting'
    this.ready = options.initialReady ?? false
    this.signal = options.signal
    this.done = new Promise((resolve) => { this.resolveDone = resolve })
    this.signal?.addEventListener('abort', this.abortListener, { once: true })
  }

  static async start(options: AgentSessionOptions): Promise<OctoberBusAgentSession> {
    if (options.signal?.aborted) throw options.signal.reason ?? new Error('Operation aborted')
    const leaseMs = options.registration.leaseMs || 300_000
    const interval = options.heartbeatIntervalMs ?? Math.floor(leaseMs / 3)
    const lifecycle = options.initialLifecycle ?? 'starting'
    const ready = options.initialReady ?? false
    if (!Number.isFinite(interval) || interval <= 0 || interval >= leaseMs) {
      throw new Error('heartbeatIntervalMs must be shorter than the execution lease')
    }
    if (lifecycle === 'offline' && ready) throw new Error('offline agents cannot be ready')
    const health = await new OctoberBusAdminClient(options.address, '').health(
      options.signal === undefined ? {} : { signal: options.signal }
    )
    if (health.name !== 'october-bus' || health.protocolVersion !== '0.1' ||
        health.status !== 'ready' || !Array.isArray(health.features) || !health.features.includes('session-retirement')) {
      throw new BusError('CONFLICT', 'Managed sessions require a ready protocol 0.1 runtime advertising session-retirement; upgrade the daemon before registering')
    }
    if (options.signal?.aborted) throw options.signal.reason ?? new Error('Operation aborted')
    const scope = new OctoberBusScopeClient(options.address, options.scopeToken)
    // Do not abort registration halfway through an ambiguous committed response.
    // Once its result arrives, cancellation retires the returned execution.
    const registration = await scope.registerAgent({ ...options.registration, leaseMs })
    const session = new OctoberBusAgentSession(options, registration)
    try {
      if (options.signal?.aborted) throw options.signal.reason ?? new Error('Operation aborted')
      await session.enqueueHeartbeat(lifecycle, ready)
      if (options.signal?.aborted) throw options.signal.reason ?? new Error('Operation aborted')
      session.scheduleHeartbeat()
      return session
    } catch (error) {
      await session.close()
      throw error
    }
  }

  get error(): unknown { return this.sessionError }

  async setState(lifecycle: AgentLifecycle, ready: boolean): Promise<Agent> {
    if (this.closed) throw new Error('agent session is closed')
    if (lifecycle === 'offline' && ready) throw new Error('offline agents cannot be ready')
    return this.enqueueHeartbeat(lifecycle, ready)
  }

  close(): Promise<void> {
    if (this.closePromise !== undefined) return this.closePromise
    this.closed = true
    this.lifecycleAbort.abort(new Error('agent session is closed'))
    if (this.timer !== undefined) clearTimeout(this.timer)
    this.signal?.removeEventListener('abort', this.abortListener)
    this.closePromise = (async () => {
      try {
        await this.lifecycleQueue
        await this.client.retire({ timeoutMs: 5_000 })
      } catch (error) {
        if (this.sessionError === undefined) this.sessionError = error
      } finally {
        this.resolveDone()
      }
    })()
    return this.closePromise
  }

  private enqueueHeartbeat(lifecycle?: AgentLifecycle, ready?: boolean): Promise<Agent> {
    const operation = this.lifecycleQueue.then(async () => {
      if (this.closed) throw new Error('agent session is closed')
      // Background heartbeats read the last confirmed state when their queued
      // turn begins, not while an earlier state write is still in flight.
      const nextLifecycle = lifecycle ?? this.lifecycle
      const nextReady = ready ?? this.ready
      const agent = await this.client.heartbeat(nextLifecycle, nextReady, this.leaseMs, { signal: this.lifecycleAbort.signal })
      this.lifecycle = nextLifecycle
      this.ready = nextReady
      return agent
    })
    // Keep the queue drainable after a failed write so cleanup is never skipped.
    this.lifecycleQueue = operation.then(() => undefined, () => undefined)
    return operation
  }

  private scheduleHeartbeat(): void {
    if (this.closed) return
    this.timer = setTimeout(() => {
      void this.enqueueHeartbeat().then(
        () => this.scheduleHeartbeat(),
        (error: unknown) => {
          // Close intentionally aborts an in-flight heartbeat; that is not a
          // failure of an otherwise clean retirement.
          if (!this.closed) this.sessionError = error
          void this.close()
        }
      )
    }, this.heartbeatIntervalMs)
  }
}

export async function* pollInbox(
  client: OctoberBusClient,
  options: InboxPollingOptions = {}
): AsyncGenerator<BusMessage[]> {
  const limit = options.limit ?? 50
  const waitMs = options.waitMs ?? 25_000
  if (limit < 1 || limit > 100) throw new Error('limit must be between 1 and 100')
  if (!Number.isInteger(waitMs) || waitMs < 1 || waitMs > 25_000) {
    throw new Error('waitMs must be an integer between 1 and 25000')
  }
  while (!options.signal?.aborted) {
    let messages: BusMessage[]
    try {
      messages = await client.pullInbox(limit, {
        waitMs,
        ...(options.signal === undefined ? {} : { signal: options.signal })
      })
    } catch (error) {
      if (options.signal?.aborted) return
      throw error
    }
    if (messages.length > 0) yield messages
  }
}

export async function withClaimedTask<T>(
  client: OctoberBusClient,
  taskId: string,
  work: (task: BusTask) => Promise<T>,
  completionNote?: (value: T) => string | undefined
): Promise<ClaimedTaskResult<T>> {
  const claimed = await client.claimTask(taskId)
  try {
    const value = await work(claimed)
    const task = await client.completeTask(taskId, completionNote?.(value))
    return { task, value }
  } catch (error) {
    try {
      await client.releaseTask(taskId)
    } catch {
      // Preserve the work error. Lease recovery remains the final fallback.
    }
    throw error
  }
}

File: bus/session.go
package bus

import (
	"context"
	"net/http"
	"slices"
	"sync"
	"time"
)

// AgentSessionOptions describes one managed agent execution.
type AgentSessionOptions struct {
	Address           string
	ScopeToken        string
	Registration      RegisterAgentInput
	HeartbeatInterval time.Duration
	InitialLifecycle  AgentLifecycle
	InitialReady      bool
	HTTP              *http.Client
}

// AgentSession keeps an agent execution registered and reachable while its
// harness process is running.
type AgentSession struct {
	Address      string
	Registration RegisterAgentResult
	Client       Client

	leaseMS   int64
	context   context.Context
	cancel    context.CancelFunc
	done      chan struct{}
	beatMu    sync.Mutex
	stateMu   sync.Mutex
	lifecycle AgentLifecycle
	ready     bool
	errMu     sync.Mutex
	err       error
	close     sync.Once
	closed    bool // guarded by stateMu
}

// StartAgentSession registers an execution and starts its heartbeat loop.
func StartAgentSession(ctx context.Context, options AgentSessionOptions) (*AgentSession, error) {
	leaseMS, err := normalizedLease(options.Registration.LeaseMS)
	if err != nil {
		return nil, err
	}
	interval := options.HeartbeatInterval
	if interval == 0 {
		interval = time.Duration(leaseMS) * time.Millisecond / 3
	}
	if interval <= 0 || interval >= time.Duration(leaseMS)*time.Millisecond {
		return nil, Errorf(CodeInvalidArgument, "heartbeat interval must be shorter than the execution lease")
	}
	lifecycle := options.InitialLifecycle
	if lifecycle == "" {
		lifecycle = LifecycleStarting
	}
	if err := validateLifecycle(lifecycle); err != nil {
		return nil, err
	}
	if lifecycle == LifecycleOffline && options.InitialReady {
		return nil, Errorf(CodeInvalidArgument, "offline agents cannot be ready")
	}
	registrationInput := options.Registration
	registrationInput.LeaseMS = leaseMS
	scopeClient := Client{Address: options.Address, Token: options.ScopeToken, HTTP: options.HTTP}
	// Check before registration: an incompatible daemon must not replace an
	// existing execution and then fail only when the new session is closed.
	health, err := scopeClient.Health(ctx)
	if err != nil {
		return nil, err
	}
	if health.Name != "october-bus" || health.ProtocolVersion != ProtocolVersion ||
		health.Status != "ready" || !slices.Contains(health.Features, FeatureSessionRetirement) {
		return nil, Errorf(CodeConflict, "managed sessions require a ready protocol 0.1 runtime advertising session-retirement; upgrade the daemon before registering")
	}
	registration, err := scopeClient.RegisterAgent(ctx, registrationInput)
	if err != nil {
		return nil, err
	}
	agentClient := Client{Address: options.Address, Token: registration.AgentToken, HTTP: options.HTTP}
	if _, err := agentClient.Heartbeat(ctx, HeartbeatInput{Lifecycle: lifecycle, Ready: options.InitialReady, LeaseMS: leaseMS}); err != nil {
		cleanupCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()
		_ = agentClient.Retire(cleanupCtx)
		return nil, err
	}
	heartbeatCtx, cancel := context.WithCancel(ctx)
	session := &AgentSession{
		Address: options.Address, Registration: registration, Client: agentClient,
		leaseMS: leaseMS, context: heartbeatCtx, cancel: cancel, done: make(chan struct{}),
		lifecycle: lifecycle, ready: options.InitialReady,
	}
	go session.heartbeat(heartbeatCtx, interval)
	return session, nil
}

func (s *AgentSession) heartbeat(ctx context.Context, interval time.Duration) {
	defer close(s.done)
	defer func() {
		s.cancel()
		s.stateMu.Lock()
		s.closed = true
		s.stateMu.Unlock()
		s.beatMu.Lock()
		defer s.beatMu.Unlock()
		cleanupCtx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()
		if err := s.Client.Retire(cleanupCtx); err != nil {
			s.errMu.Lock()
			if s.err == nil {
				s.err = err
			}
			s.errMu.Unlock()
		}
	}()
	ticker := time.NewTicker(interval)
	defer ticker.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			if _, err := s.sendHeartbeat(ctx); err != nil {
				if ctx.Err() != nil {
					return
				}
				s.errMu.Lock()
				s.err = err
				s.errMu.Unlock()
				return
			}
		}
	}
}

func (s *AgentSession) sendHeartbeat(ctx context.Context) (Agent, error) {
	s.beatMu.Lock()
	defer s.beatMu.Unlock()
	s.stateMu.Lock()
	if s.closed {
		s.stateMu.Unlock()
		return Agent{}, Errorf(CodeConflict, "Agent session is closed")
	}
	lifecycle, ready := s.lifecycle, s.ready
	s.stateMu.Unlock()
	return s.Client.Heartbeat(ctx, HeartbeatInput{Lifecycle: lifecycle, Ready: ready, LeaseMS: s.leaseMS})
}

// SetState updates the lifecycle and readiness reported by this execution.
func (s *AgentSession) SetState(ctx context.Context, lifecycle AgentLifecycle, ready bool) (Agent, error) {
	s.stateMu.Lock()
	closed := s.closed
	s.stateMu.Unlock()
	if closed {
		return Agent{}, Errorf(CodeConflict, "Agent session is closed")
	}
	if err := validateLifecycle(lifecycle); err != nil {
		return Agent{}, err
	}
	if lifecycle == LifecycleOffline && ready {
		return Agent{}, Errorf(CodeInvalidArgument, "offline agents cannot be ready")
	}
	s.beatMu.Lock()
	defer s.beatMu.Unlock()
	s.stateMu.Lock()
	if s.closed {
		s.stateMu.Unlock()
		return Agent{}, Errorf(CodeConflict, "Agent session is closed")
	}
	s.stateMu.Unlock()
	operationContext, cancel := context.WithCancel(ctx)
	stop := context.AfterFunc(s.context, cancel)
	defer stop()
	defer cancel()
	agent, err := s.Client.Heartbeat(operationContext, HeartbeatInput{Lifecycle: lifecycle, Ready: ready, LeaseMS: s.leaseMS})
	if err == nil {
		s.stateMu.Lock()
		s.lifecycle, s.ready = lifecycle, ready
		s.stateMu.Unlock()
	}
	return agent, err
}

// Done closes after heartbeat termination and the bounded retirement attempt.
func (s *AgentSession) Done() <-chan struct{} { return s.done }

// Err returns the first heartbeat or retirement failure, if any.
func (s *AgentSession) Err() error {
	s.errMu.Lock()
	defer s.errMu.Unlock()
	return s.err
}

// Close stops heartbeats, retires authority and releases execution obligations.
func (s *AgentSession) Close(ctx context.Context) error {
	s.close.Do(func() {
		s.stateMu.Lock()
		s.closed = true
		s.stateMu.Unlock()
		s.cancel()
	})
	select {
	case <-s.done:
		return s.Err()
	case <-ctx.Done():
		return ctx.Err()
	}
}
